### PR TITLE
fix(appmesh): TLS mode is set on the Certificate class

### PR DESCRIPTION
**Pull request by Seiya6329**
This is the second part of the series for [issue#12733](https://github.com/aws/aws-cdk/issues/12733).
For the breakdown on series, please refer to
[this comment](https://github.com/aws/aws-cdk/pull/14782#discussion_r636334863).

#### Collaborators

@alexbrjo and @dfezzie. Thank you!

#### Design notes

- Adding `TlsListener` interface above `TlsCertificate` in order to use `TlsCertificate` for `TlsClientPolicy` ([example](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-clientpolicytls.html)) 
- Updating TLS certificate abstraction by removing `tlsMode` property since `tlsMode` is part `TlsListener` property. In addition, this update is also required because `TlsClientPolicy` does NOT have `tlsMode` property.

BREAKING CHANGE: the creation property `tlsCertificate` in `VirtualNode` has been renamed to `tls`, and its type changed to `TlsListener`
- **appmesh**: the creation property `tlsCertificate` in `VirtualGatewayListener` has been renamed to `tls`, and its type changed to `TlsListener`
- **appmesh**: the `tlsMode` property has been removed from the options when creating a `TlsCertificate`, moved to the new `TlsListener` interface, and renamed `mode`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by skinny85 — DISMISSED** [review]
Looks great @Seiya6329!

**Comment by Seiya6329** [comment]
### REV2:
- Separating `TlsListener` and `TlsCertificate` into their own files.

**Comment by skinny85** [comment]
@Seiya6329 can you fix the build for this PR? It's failing with

```
@aws-cdk/aws-appmesh: [96mtest/test.virtual-node.ts[0m:[93m774[0m:[93m9[0m - [91merror[0m[90m TS2345: [0mArgument of type '{ port: number; tlsCertificate: appmesh.TlsCertificate; }' is not assignable to parameter of type 'HttpVirtualNodeListenerOptions'.
@aws-cdk/aws-appmesh:   Object literal may only specify known properties, and 'tlsCertificate' does not exist in type 'HttpVirtualNodeListenerOptions'.
@aws-cdk/aws-appmesh: [7m774[0m         tlsCertificate: appmesh.TlsCertificate.file({
@aws-cdk/aws-appmesh: [7m   [0m [91m        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
@aws-cdk/aws-appmesh: [7m775[0m           certificateChainPath: 'path/to/certChain',
@aws-cdk/aws-appmesh: [7m   [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
@aws-cdk/aws-appmesh: [7m...[0m 
@aws-cdk/aws-appmesh: [7m777[0m           tlsMode: appmesh.TlsMode.PERMISSIVE,
@aws-cdk/aws-appmesh: [7m   [0m [91m~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
@aws-cdk/aws-appmesh: [7m778[0m         }),
@aws-cdk/aws-appmesh: [7m   [0m [91m~~~~~~~~~~[0m
@aws-cdk/aws-appmesh: [96mtest/test.virtual-node.ts[0m:[93m777[0m:[93m11[0m - [91merror[0m[90m TS2345: [0mArgument of type '{ certificateChainPath: string; privateKeyPath: string; tlsMode: appmesh.TlsMode; }' is not assignable to parameter of type 'FileCertificateOptions'.
@aws-cdk/aws-appmesh:   Object literal may only specify known properties, and 'tlsMode' does not exist in type 'FileCertificateOptions'.
```

**Comment by Seiya6329** [comment]
@skinny85 - I apologize for leaving the building failure.

### REV3:
- Addressing the build failure

### Test:
- `yarn run build` and `yarn run test` successful

File: packages/@aws-cdk/aws-appmesh/lib/index.ts [hunk 1]
```diff
@@ -17,3 +17,5 @@ export * from './gateway-route';
 export * from './gateway-route-spec';
 export * from './client-policy';
 export * from './health-checks';
+export { TlsListener } from './tls-listener';
+export { TlsMode } from './tls-listener';
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/index.ts:21]: Can you replace this with

```suggestion
export * from './tls-listener';
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import * as iam from '@aws-cdk/aws-iam';
 import * as cdk from '@aws-cdk/core';
 import { Test } from 'nodeunit';
 import * as appmesh from '../lib';
+import { TlsMode } from '../lib/tls-listener';
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:7]: Nope! Please remove this.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 3]
```diff
@@ -183,7 +186,7 @@ export = {
           Listeners: [
             {
               TLS: {
-                Mode: appmesh.TlsMode.STRICT,
+                Mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:189]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 4]
```diff
@@ -170,10 +171,12 @@ export = {
         mesh: mesh,
         listeners: [appmesh.VirtualGatewayListener.http({
           port: 8080,
-          tlsCertificate: appmesh.TlsCertificate.acm({
-            tlsMode: appmesh.TlsMode.STRICT,
-            certificate: cert,
-          }),
+          tls: {
+            mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:175]:
```suggestion
            mode: appmesh.TlsMode.STRICT,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 5]
```diff
@@ -214,11 +217,13 @@ export = {
         mesh: mesh,
         listeners: [appmesh.VirtualGatewayListener.grpc({
           port: 8080,
-          tlsCertificate: appmesh.TlsCertificate.file({
-            certificateChainPath: 'path/to/certChain',
-            privateKeyPath: 'path/to/privateKey',
-            tlsMode: appmesh.TlsMode.STRICT,
-          }),
+          tls: {
+            mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:221]:
```suggestion
            mode: appmesh.TlsMode.STRICT,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 6]
```diff
@@ -228,7 +233,7 @@ export = {
           Listeners: [
             {
               TLS: {
-                Mode: appmesh.TlsMode.STRICT,
+                Mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:236]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 7]
```diff
@@ -258,11 +263,13 @@ export = {
         mesh: mesh,
         listeners: [appmesh.VirtualGatewayListener.grpc({
           port: 8080,
-          tlsCertificate: appmesh.TlsCertificate.file({
-            certificateChainPath: 'path/to/certChain',
-            privateKeyPath: 'path/to/privateKey',
-            tlsMode: appmesh.TlsMode.PERMISSIVE,
-          }),
+          tls: {
+            mode: TlsMode.PERMISSIVE,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:267]:
```suggestion
            mode: appmesh.TlsMode.PERMISSIVE,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 8]
```diff
@@ -272,7 +279,7 @@ export = {
           Listeners: [
             {
               TLS: {
-                Mode: appmesh.TlsMode.PERMISSIVE,
+                Mode: TlsMode.PERMISSIVE,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:282]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 9]
```diff
@@ -5,6 +5,7 @@ import * as iam from '@aws-cdk/aws-iam';
 import * as cdk from '@aws-cdk/core';
 import { Test } from 'nodeunit';
 import * as appmesh from '../lib';
+import { TlsMode } from '../lib/tls-listener';
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:8]: Nope! This line should not be added.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 10]
```diff
@@ -438,22 +439,23 @@ export = {
           mesh,
           listeners: [appmesh.VirtualNodeListener.grpc({
             port: 80,
-            tlsCertificate: appmesh.TlsCertificate.acm({
-              certificate: cert,
-              tlsMode: appmesh.TlsMode.STRICT,
-            }),
+            tls: {
+              mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:443]:
```suggestion
              mode: appmesh.TlsMode.STRICT,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 11]
```diff
@@ -438,22 +439,23 @@ export = {
...
 
         // THEN
-
         expect(stack).to(haveResourceLike('AWS::AppMesh::VirtualNode', {
           Spec: {
             Listeners: [
               {
                 TLS: {
-                  Mode: appmesh.TlsMode.STRICT,
+                  Mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:458]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 12]
```diff
@@ -438,22 +439,23 @@ export = {
...
+              certificate: appmesh.TlsCertificate.acm({
+                certificate: cert,
+              }),
+            },
           },
           )],
         });
 
         // THEN
-
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:450]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 13]
```diff
@@ -485,11 +487,13 @@ export = {
           mesh,
           listeners: [appmesh.VirtualNodeListener.http({
             port: 80,
-            tlsCertificate: appmesh.TlsCertificate.file({
-              certificateChainPath: 'path/to/certChain',
-              privateKeyPath: 'path/to/privateKey',
-              tlsMode: appmesh.TlsMode.STRICT,
-            }),
+            tls: {
+              mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:491]:
```suggestion
              mode: appmesh.TlsMode.STRICT,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 14]
```diff
@@ -499,7 +503,7 @@ export = {
             Listeners: [
               {
                 TLS: {
-                  Mode: appmesh.TlsMode.STRICT,
+                  Mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:506]: This should not be changed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 15]
```diff
@@ -530,11 +534,13 @@ export = {
           mesh,
           listeners: [appmesh.VirtualNodeListener.http({
             port: 80,
-            tlsCertificate: appmesh.TlsCertificate.file({
-              certificateChainPath: 'path/to/certChain',
-              privateKeyPath: 'path/to/privateKey',
-              tlsMode: appmesh.TlsMode.PERMISSIVE,
-            }),
+            tls: {
+              mode: TlsMode.PERMISSIVE,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:538]:
```suggestion
              mode: appmesh.TlsMode.PERMISSIVE,
```

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 16]
```diff
@@ -544,7 +550,7 @@ export = {
             Listeners: [
               {
                 TLS: {
-                  Mode: appmesh.TlsMode.PERMISSIVE,
+                  Mode: TlsMode.PERMISSIVE,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:553]: This should not have been changed.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Almost there! Some weird changes sneaked into the tests.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts [hunk 17]
```diff
@@ -5,7 +5,8 @@ import {
   GrpcConnectionPool, GrpcTimeout, Http2ConnectionPool, HttpConnectionPool,
   HttpTimeout, OutlierDetection, Protocol, TcpConnectionPool, TcpTimeout,
 } from './shared-interfaces';
-import { TlsCertificate, TlsCertificateConfig } from './tls-certificate';
+import { TlsListener } from './tls-listener';
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:9]: No need for this extra empty line here.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts [hunk 18]
```diff
@@ -7,7 +7,8 @@ import {
   HttpConnectionPool,
   Protocol,
 } from './shared-interfaces';
-import { TlsCertificate, TlsCertificateConfig } from './tls-certificate';
+import { TlsListener } from './tls-listener';
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts:11]: No need for this extra empty line.

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 19]
```diff
@@ -248,13 +248,15 @@ const cert = new certificatemanager.Certificate(this, 'cert', {...});
...
-  dnsHostName: 'node',
+  serviceDiscovery: appmesh.ServiceDiscovery.dns('node'),
   listeners: [appmesh.VirtualNodeListener.grpc({
     port: 80,
-    tlsCertificate: appmesh.TlsCertificate.acm({
-      certificate: cert,
-      tlsMode: TlsMode.STRICT,
-    }),
+    tls: {
+      mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:255]:
```suggestion
      mode: apmesh.TlsMode.STRICT,
````

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 20]
```diff
@@ -263,11 +265,13 @@ const gateway = new appmesh.VirtualGateway(this, 'gateway', {
   mesh: mesh,
   listeners: [appmesh.VirtualGatewayListener.grpc({
     port: 8080,
-    tlsCertificate: appmesh.TlsCertificate.file({
-      certificateChain: 'path/to/certChain',
-      privateKey: 'path/to/privateKey',
-      tlsMode: TlsMode.STRICT,
-    }),
+    tls: {
+      mode: TlsMode.STRICT,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:269]:
```suggestion
      mode: apmesh.TlsMode.STRICT,
````

File: packages/@aws-cdk/aws-appmesh/lib/index.ts (hunk 1 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/index.ts:21]: Addressed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts (hunk 2 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:7]: Removed.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts (hunk 3 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:189]: Reverted

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts (hunk 12 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:450]: This original spacing is odd and other locations don't have spacing after `// Then` comment. Although it is not specifically related to this code change, I would like to make a change for it.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts (hunk 17 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts:9]: Removing the extra line.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts (hunk 18 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-gateway-listener.ts:11]: Removed

**Comment by Seiya6329** [comment]
### REV4:
- Fixing the files the way to import `TlsMode`

### NOTE:
- Remaining the line deletion ([Ref](https://github.com/aws/aws-cdk/pull/14856#discussion_r640846550))

### TEST:
- `yarn run build` and `yarn run test` successful.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts (hunk 12 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:450]: OK.

**Review by skinny85 — APPROVED** [review]
Looks great @Seiya6329!

**Pull request merged**